### ISSUE TITLE
Synctex / forward search crashes

### ISSUE THREAD

**Issue by lunbudao** [sumatrapdfreader/sumatrapdf#3381]
### Discussed in https://github.com/sumatrapdfreader/sumatrapdf/discussions/3380

<div type='discussions-op-text'>

<sup>Originally posted by **kjk** March 25, 2023</sup>
Thanks to efforts of @avorobey In latest [pre-release version](https://www.sumatrapdfreader.org/prerelease) you can:

* move tabs between window
* drag tabs within window
* drag out a tab to create a new window.

Also tweaked the display of tabs.for bigger contrast between tab and background, so it's easier to tell selected tab.
</div>

**Comment by lunbudao** [comment]
I love this feature. Current issue is it become easily crash (if not always) when using forward-search of latex (the inverse-search work well), my tool combination is vscode + latex workshop + miktex + sumatrapdf (build 15374), and the 3.4.6 version work well with both forward-search and inverse-search.

**Comment by GitHubRulesOK** [comment]
Will need much more details so follow the crash report guidance on webpage, however latex and vscode as odd bedfellows hence the extension. Latex comes with its own IDE and MiKTeX is not static like Tex Live. 
Debugging is thus often working through many variables. I suggest best to test SumatraPDF against the Miktex editor TexWorks?? to check if it crashes without VSCode. 

All the normal rules apply (avoid compressed synctex) check forward search is not  constantly reconfiguring just calling SumatraPDF as view only and naturally that it was a run free compile.

**Comment by lunbudao** [comment]
I compile a very simple tex file using command "pdflatex.exe [filename] -synctex=-1"  to get an uncompressed synctex. but I can't using forward search with sumatrapdf in TexWorks. So I have to use vscode again, sorry for that, still find that the build 15374 version sumatrapdf do not work with forward search (not responding and then crash), and 3.4.6 version sumatrapdf work well.

I think I can not provide more information for this problem!

**Comment by avorobey** [comment]
@lunbudao, I want to look at and solve the bug, and I'm willing to install vscode, but I've never used "forward search" and don't fully understand how it works. Do you think you can 
- upload here the very simple tex file *and* the compiled pdf file (then I don't need to install pdflatex, right?)
- explain what I need to do in vscode to trigger the crash: if I need to configure it to use SumatraPDF in any way, beyond installing it, or what to do to cause the crash.

Thanks for your help already!

**Comment by GitHubRulesOK** [comment]
@avorobey
Thanks very much perhaps this overview may help 
https://github.com/GitHubRulesOK/MyNotes/raw/master/AppNotes/SumatraPDF/LATeX%20and%20Reverse-Search.zip
 should have sample test and include a  copy of 
https://github.com/GitHubRulesOK/MyNotes/raw/master/AppNotes/SumatraPDF/LATeX%20and%20Reverse-Search.pdf beware there have been resent changes (past 2-3 years for Workshop that make the guide outof date for that case, but principles are similar) thus much would depend on configuration but if its forward not inverse there is little to test in SumatraPDF handling.

The sample files are a good way to test reverse search as a fixed set butt unsure if big enough to trigger a forwards fail where often its the editor compile that's the cause

**Comment by GitHubRulesOK** [comment]
@avorobey 
TL;dr thats a slow way to learn your target is the interaction as shown there by

```
Microsoft VS Code = "\editor\Code.exe” -g %f:%l [you may need to try "%f:%l" or other escape chars]
```
however that was possibly changed  to dde calling

the sample file is set as follows
![image](https://user-images.githubusercontent.com/12754913/227715333-29e0c598-aaba-4090-94fb-9dad431394ff.png)

**Comment by lunbudao** [comment]
@avorobey I am very appreciated for that. Actually, I'm not very sure whether you need to install latex (I use Miktex distrubution) even if I offer these files. You can have a try (at the worst condition, you may need to install  Miktex, 135 MB installer, or so). 

[helloworld.zip](https://github.com/sumatrapdfreader/sumatrapdf/files/11069282/helloworld.zip)

To trigger the crash, you need first configure vscode and sumatrapdf to implement forward search, the steps are:

1. download vscode, the install "latex workshop" extension from the extension marketplace of vscode (easy to do that).
2. download sumatrapdf (build version 15374 or 15375, I have try this two, both cause crash), I think newer version (>15375) is also ok.
3. Then, in vscode, open the settings.json, paste the content of "latex-workshop_settings.jsonc", which is in the compressed file, into settings.json. You NEED to adjust the "Path/to/SumatraPDF.exe" strings (two strings in that file) to your own path. Then you should be able to use forward search (if not, you may need to install Miktex or so), just open the folder where you put the tex file (pdf file, and synctex file) in vscode, open tex file
4. THEN, click on some place in tex file (make sure you are in a state that if you hit the keyboard, the tex file content changed in the editor), and then press the shortcut "ctrl+alt+J", it should automatically open the sumatrapdf.
5. AND THEN, it will CRASH when you repeat step 4 again.

I hope this help.

ADDITIONAL INFORMATION: I use lualatex to compile tex file, and use argument "-synctex=-1" here (which means not to compress synctex file), the default behavior is "-synctex=1" (which means to compress synctex file), I change that because it solve the problem that inverse search fail. But I think both are ok to cause the crash.
![image](https://user-images.githubusercontent.com/24380551/227728110-665bb224-1522-4100-9cdf-e1febadc606d.png)

**Comment by GitHubRulesOK** [comment]
@avorobey 
Not using VSCode I am at  a disadvantage as to why it needs to be complicated

from guide
---
```
"latex-workshop.view.pdf.viewer": "external",
"latex-workshop.view.pdf.external.synctex.command": "C:/Users/zhang/AppData/Local/SumatraPDF/SumatraPDF.exe",
  "latex-workshop.view.pdf.external.synctex.args": [
    "-forward-search",
    "%TEX%",
    "%LINE%",
    "-reuse-instance",
    "-inverse-search",
    "\"C:\\Users\\<Username>\\AppData\\Local\\Programs\\Microsoft VS Code\\Code.exe\" \"C:\\Users\\<U>\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\cli.js\" --ms-enable-electron-run-as-node -r -g \"%f:%l\"",
    "%PDF%"
  ],
```
the ***-reuse-instance*** should NOT BE REQUIRED (more likely to hicup)
the ***-inverse search*** is NOT REQUIRED (more likely to bork a SumatraPDF installation)

**the command should simply be "SumatraPDF.exe" -forward-search %TEX% %LINE%**
---

**Comment by kjk** [comment]
To provide some info.

Debugging this is difficult because reverse search launches a new process which then sends commands over DDE to existing process, so you have to set the debugging. See `SearchAndDDE.cpp`.

The code is `PdfSync.cpp` and mostly uses third-party `synctex_parsre.c` which is the worst code ever.

Also we watch changes to synctex file and re-load it ourselves, so the tools over-write synctex and we pick up those changes automatically.

Synctex is a file with mapping info between position in source latex files and generated PDF, which allows jumping from PDF in SumatraPDF directly to latex code that generated it.

Even though I didn't make changes (that I remember) to PdfSync.cpp or synctex_parser.c, users started reporting elevated rates of PdfSync failing (but not crashes).

By adding logging I've tracked it down to synctex  seemingly failing to load the `synctex.gz` file even though I could read it just fine in PdfSync.cpp.

I theorized antivirus temporarily blocking but that doesn't seem so.

One change I made was switch from `zlib` to `zlib-ng` (supposed to be faster) which I reverted because I was seeing crashes from it. I just noticed there was a remaining issue from that which I just fixed https://github.com/sumatrapdfreader/sumatrapdf/commit/0f6c4107824ec245551b40f5f5081590edac2781

I contemplated re-writing syntex parser code, because it's the worst but it's rather big.

It can't even be reasonably re-factored. I wanted to move reading the contest of synctex file to my code and have synctex-parser.c do parsing from memory but the code just reads from file descriptor and it's scattered all over the code. Like I said: the worst code ever.

You can also see: https://github.com/sumatrapdfreader/sumatrapdf/issues/2262 and https://github.com/sumatrapdfreader/sumatrapdf/issues/2642 and https://github.com/sumatrapdfreader/sumatrapdf/discussions/2640 and history of `PdfSync.cpp` to see my changes trying to debug this.

**Comment by kjk** [comment]
@GitHubRulesOK what does `crash` means? I'm not seeing crashes.

**Comment by GitHubRulesOK** [comment]
@kjk @avorobey 
The crash in portable only seem to be most recent so this seems to trigger on second call

"C:\Apps\PDF\SumatraPDF\3.5\SumatraPDF-15375-64.exe" -forward-search sync.tex 21 sync.pdf

**Comment by GitHubRulesOK** [comment]
Ok this is odd if I delete history its now not happening it was on second call, but had other sessions in background due to other issue.??

**Comment by kjk** [comment]
Also, do people experiencing syntex issue use OneDrive or some other networked file system to store the synctex files?

**Comment by kjk** [comment]
Related: it seems synctex doesn't open non-english file paths.

One change I did do was refactoring of PdfSync.cpp to use WCHAR* less and synctex is peculiar about synctex_scanner_new_with_output_file path.

I could have change some subtle thing.

Then again those issues are not consistent so maybe not.

**Comment by GitHubRulesOK** [comment]
these test files so old forgot when built but usually a good indicator of a normal run, am using standard drive the only common factor seems to be more than one sumatrapdf active.

**Comment by kjk** [comment]
Could people re-try with latest https://www.sumatrapdfreader.org/prerelease 15376 or later?

I just made a change that could actually fix it.

There was a mismatch of header vs. code i.e. I was compiling zlib but using zlib-ng headers.

That doesn't affect most of the code because they are supposed to be API compatible but synctex is the only code (I thing) that uses `gzopen` etc. functions so if the definitions of FILE* differ, this could explain why I could read the .synctex.gz file in Sumatra code but synctex would fail to load it and even crash.

**Comment by GitHubRulesOK** [comment]
crashes with 2 sessions active but I was pushing it by using 2 versions
![image](https://user-images.githubusercontent.com/12754913/227739039-6789d963-3a1e-45ba-a53a-c356b98c78f9.png)
One session 2 tabs 2nd call still causes crash
![image](https://user-images.githubusercontent.com/12754913/227739100-3c95f3f4-8e96-4b13-b614-db827b61f26d.png)

**Comment by GitHubRulesOK** [comment]
On its own this is hard to trace here one session 2 tabs many calls no fail
![image](https://user-images.githubusercontent.com/12754913/227739214-ab4ad8a2-a68c-4edc-837e-7e773b4b4415.png)

**Comment by GitHubRulesOK** [comment]
@lunbudao 
Can you please try
[pre-release 3.5.15380 +](https://www.sumatrapdfreader.org/prerelease)

**Comment by GitHubRulesOK** [comment]
@lunbudao 
first you need to ensure the runtime log shows that the files were built as a set correctly at runtime (thats why I use my simple test files they were a perfect run even if a minimal set) you can use a tex file as small as fails in tour system.
then supply the tex synctex and pdf (they should all 3 have exactly the same date and time in explorer, thus show as a work set) you can put all 3 in azip and paste here in this thread.

**Comment by GitHubRulesOK** [comment]
@lunbudao 

Looking at your supplied files I see in the Json configuration
```
"latex-workshop.view.pdf.external.synctex.command": "Path/to/SumatraPDF.exe",
"latex-workshop.view.pdf.external.synctex.args": [
    "-forward-search",
    "%TEX%",
    "%LINE%",
    "%PDF%",
```

I would have expected `Path/to/SumatraPDF.exe` to be something like `C:/Program Files/SumatraPDF/SumatraPDF.exe`
ALSO the last line of a json section is usually without a separator `,` e.g. `   just  `"%PDF%"`

**Comment by GitHubRulesOK** [comment]
@lunbudao @avorobey  @kjk 

That last comment seems to be the trigger a malformed call will cause a crash

![image](https://user-images.githubusercontent.com/12754913/228386225-c7014951-61d6-4498-9504-d955504048e3.png)

**Comment by kjk** [comment]
Does it generate a crash report on your machine? Could you attach it?

**Comment by kjk** [comment]
Never mind, I can reproduce the crash. Will fix.

**Comment by lunbudao** [comment]
> @lunbudao
> 
> Looking at your supplied files I see in the Json configuration
> 
> ```
> "latex-workshop.view.pdf.external.synctex.command": "Path/to/SumatraPDF.exe",
> "latex-workshop.view.pdf.external.synctex.args": [
>     "-forward-search",
>     "%TEX%",
>     "%LINE%",
>     "%PDF%",
> ```
> 
> I would have expected `Path/to/SumatraPDF.exe` to be something like `"C:/Program Files/SumatraPDF/SumatraPDF.exe"` or an appdata location ALSO the last line of a json section is usually without a separator `,` e.g. just `"%PDF%"`

Yes, the "Path/to/Sumatrapdf.exe" is just like that.
And about the comma after last line, I think it was a problem in vscode, but after an update of vscode (I think it was years ago), this usage seems work well.

**Comment by kjk** [comment]
I was able to reproduce a fix a crash when calling sumatra like `-forward-search "C:\Users\kjk\Downloads\helloworld\helloworld.tex" 3  "C:\Users\kjk\Downloads\helloworld\helloworld.pdf"`

So hopefully your crash is fixed as well.

It was caused by memory corruption due to refactoring for temp allocator.

I wasn't seeing those crashes because it seem crashes caused by memory corruption won't generate crash reports.

**Comment by lunbudao** [comment]
@kjk Do you fix it in 15386 build version, I still suffer the problem with 15386.

**Comment by kjk** [comment]
It was fixed in 15388

**Comment by lunbudao** [comment]
@kjk Thanks very much for your hard work, it works now. @avorobey @GitHubRulesOK Thanks for your efforts, I'm very appreciated.
